### ISSUE TITLE
-H:+AllowVMInspection results in java.lang.NoSuchMethodException: jdk.vm.ci.hotspot.HotSpotResolvedJavaMethodImpl.getMetaspaceMethod()

### ISSUE THREAD

**Issue by zakkak** [graalvm/mandrel#262]
## Description
Building a native image with `-H:+AllowVMInspection` fails with:
```
$ native-image -H:+AllowVMInspection HelloWorld                            
[helloworld:337555]    classlist:   1,193.54 ms,  0.96 GB
[helloworld:337555]        (cap):     447.76 ms,  0.96 GB
[helloworld:337555]        setup:   1,205.43 ms,  0.96 GB
Fatal error:com.oracle.svm.core.util.VMError$HostedError: java.lang.NoSuchMethodException: jdk.vm.ci.hotspot.HotSpotResolvedJavaMethodImpl.getMetaspaceMethod()
	at com.oracle.svm.core.util.VMError.shouldNotReachHere(VMError.java:72)
	at com.oracle.svm.jfr.JfrEventSubstitution.setPublicModifier(JfrEventSubstitution.java:123)
	at com.oracle.svm.jfr.JfrEventSubstitution.changeWriterResetMethod(JfrEventSubstitution.java:98)
	at com.oracle.svm.jfr.JfrEventSubstitution.<init>(JfrEventSubstitution.java:63)
	at com.oracle.svm.jfr.JfrFeature.duringSetup(JfrFeature.java:138)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$setupNativeImage$18(NativeImageGenerator.java:928)
	at com.oracle.svm.hosted.FeatureHandler.forEachFeature(FeatureHandler.java:71)
	at com.oracle.svm.hosted.NativeImageGenerator.setupNativeImage(NativeImageGenerator.java:928)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:571)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$2(NativeImageGenerator.java:497)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1407)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.lang.NoSuchMethodException: jdk.vm.ci.hotspot.HotSpotResolvedJavaMethodImpl.getMetaspaceMethod()
	at java.base/java.lang.Class.getDeclaredMethod(Class.java:2475)
	at com.oracle.svm.jfr.JfrEventSubstitution.setPublicModifier(JfrEventSubstitution.java:106)
	... 14 more
[helloworld:337555]      [total]:   2,443.49 ms,  0.96 GB
# Printing build artifacts to: /home/zakkak/code/mandrel-packaging/helloworld.build_artifacts.txt
Error: Image build request failed with exit status 1
```

## How To Reproduce
1. `native-image -H:+AllowVMInspection HelloWorld`

### Configuration:
 - OS: Fedora 34
 - Architecture: AMD64
 - Mandrel version: 21.2.0-devc03bfc33d5.tar.gz
 - JDK version: 11.0.12-ea+5

## Expected behavior
Compilation should not fail

**Comment by jiekang** [comment]
For what it's worth, I found that `getMetaspaceMethod()` was added to openjdk latest here:

https://github.com/openjdk/jdk/commit/e9c523ae5f501d61b0bb7a8e725b3d5de7f3423e#diff-b39038af11e0ac6774aa1b7e66c1b11fa265194649762a4768446f74d6fbf2cdR211

https://bugs.openjdk.java.net/browse/JDK-8220623

**Comment by jerboaa** [comment]
`getMetaspaceMethod()` was introduced in OpenJDK with [JDK-8220623](https://bugs.openjdk.java.net/browse/JDK-8220623), a JDK 13 change which hasn't been backported to OpenJDK 11u. Perhaps `getMetaspacepointer()` could be used instead. It would entail some reflection...

**Comment by jerboaa** [comment]
So this has to do with a different JVMCI level. Graal VM CE would likely not run into this as they are on a more recent JVMCI level.

**Comment by jerboaa** [comment]
This should fix it:

```
diff --git a/substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrEventSubstitution.java b/substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrEventSubstitution.java
index b8528a34dac..c659929dc9e 100644
--- a/substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrEventSubstitution.java
+++ b/substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrEventSubstitution.java
@@ -103,7 +103,15 @@ public class JfrEventSubstitution extends SubstitutionProcessor {
     private static void setPublicModifier(ResolvedJavaMethod m) {
         try {
             Class<?> method = m.getClass();
-            Method metaspaceMethodM = method.getDeclaredMethod("getMetaspaceMethod");
+            Method metaspaceMethodM = null;
+            try {
+                metaspaceMethodM = method.getDeclaredMethod("getMetaspaceMethod");
+            } catch (NoSuchMethodException e) {
+                // Fall back for OpenJDK 11u-based builds:
+                //  getMetaspaceMethod() got introduced into OpenJDK 13 with JDK-8220623.
+                //  Plain OpenJDK 11u-based code won't have it. Use getMetaspacePointer() instead.
+                metaspaceMethodM = method.getDeclaredMethod("getMetaspacePointer");
+            }
             metaspaceMethodM.setAccessible(true);
             long metaspaceMethod = (Long) metaspaceMethodM.invoke(m);
             // Checkstyle: stop

```

**Comment by jerboaa** [comment]
I'll try to get this into upstream Graal VM.

**Comment by jerboaa** [comment]
Upstream PR: https://github.com/oracle/graal/pull/3488

**Comment by jerboaa** [comment]
I'm not sure how this can affect 21.1. It didn't have the JFR feature.

**Comment by zakkak** [comment]
This is now merged in upstream `master`, pending backport to `release/graal-vm/21.2`.

**Comment by zakkak** [comment]
Fixed by 6c6bdbd470d6297fe1c8d833b26cf9e75b32f900 and f4e3191a5e5351d25e082d4a0df825084e05c6eb
